### ISSUE TITLE
kernel handling of hwrng versus rngd

### ISSUE THREAD

**Issue by dermotbradley** [nhorman/rng-tools#167]
In recent Linux kernels it appears that the kernel itself will use any HWRNG present to mix entropy into the pool. It is not clear however if such entropy will be credited (it depends on the "quality" value defined by the specific kernel module).

I'm wondering what impact this may have on rngd (i.e. if both the kernel and rngd are using the same HWRNG to mix in entropy) and also, if the kernel is indeed crediting the entropy, whether this in some cases removes the need for run rngd (unless other additional entropy sources such as jitter are required).

**Comment by nhorman** [comment]
I'm pretty sure that no one has done the math on that, nor can anyone really, unless the specific hwrng in use is known (i.e. how much entropy can it produce per unit time, what its measured entropy is, etc).  the exact behavior would rather be an artifact of those, and other aspects of the hardware).

I also don't think you're correct in your initial assertion.  I'm looking at the /dev/random driver right now and, as per usual, the only calls to write_pool in it are from the /dev/random ioctl RNDADDENTROPY and from the character drivers write method (both of which only come from user space).  All the rest of the entropy is generated by a crng that is periodically reseeded.  There might be something hanging out in some pending branch somewhere, but I'm unaware of it.

So I suppose as to the crediting question, the answer is no (because the obtaining of entropy via that method doesn't exist as far as I can see)

the one exception to the case is during rng initialization on boot, in which the entropy pool is initially seeded.  That (on a per arch basis), may come from a hardware rng, if available (which must be supported by the arch ISA, for instance the RDRAND instruction on x8664 or the RNDR instruction on some arm platforms).  If those instructions don't exist though seeing is done with whatever stack data happens to be there during init, mixed with a few other pieces of data (a time call and utsname call IIRC).

**Comment by dermotbradley** [comment]
> That (on a per arch basis), may come from a hardware rng, if available (which must be supported by the arch ISA, for instance the RDRAND instruction on x8664 or the RNDR instruction on some arm platforms). If those instructions don't exist though seeing is done with whatever stack data happens to be there during init, mixed with a few other pieces of data (a time call and utsname call IIRC).

I was thinking of the following:

https://github.com/torvalds/linux/blob/master/drivers/char/random.c#L932-L934

https://github.com/torvalds/linux/blob/master/drivers/char/random.c#L1135-L1162

that function is called by https://github.com/torvalds/linux/blob/master/drivers/char/hw_random/core.c#L491-L534

**Comment by nhorman** [comment]
ah, I stand corrected. looks like that was added pretty recently (in Feb it appears).

So, given that, looking at that function, it appears that the entropy collected from a hwrng is in fact credited (you can see it in the last function called by add_hwgenerator_randomness).  There we cal credit_entropy_bit, which credits the /dev/random pool to the minimum of the amount of data read from the hwrng or the maximum pool size.  The quality value for the specific hwrng in used is used as a multiplicative factor in hwrng_fillfn to determine the amount of entropy provided in one read iteration (measured in 1/1024'ths of a bit

As to your question regarding behavior of intermixing the use of rngd and this mechanism, I think the answer is still non-deterministc.  The expectation is that a hwrng is a universal source of entropy (i.e. never running out of bits to give), and as such there should be no problem in running both in parallel (the kernel and rngd should be able to read from the hwrng in parallel, getting separate streams of entropy which will then be mixed together).  Practically speaking though, that may not be the case and behavior will be based on the specific hwrng in use.  For instance, if you're using ppc64s DARN instruction, and read to fast from it, you may get errors in reading from the instruction in the kernel or rngd, as the DARN instruction will return errors when the hardware doesn't feel it can provide sufficient entropy.  RDRAND on x86_64 tends to be less cautious about that sort of thing, and will likely not be problematic.  If you're using an AMD64 Ryzen 3000 with the rdrand bug, you will likely get all zeros and no error in both streams, and will get a very bad entropy pool (which granted, is a hardware problem, and outside the scope of this question, but you get the idea).

Long story short, it shouldn't be a problem to do reads in both places, and proper error handling is implement in both the kernel and rngd, so it should be fine, but you always want to know what your hardware is doing, and be aware of any shortcomings so you can take corrective action.
